gh-101944: mistake in documentation for `PyModule_AddObjectRef()` (#101957)
python/cpython: Doc/c-api/module.rst

   Add an object to *module* as *name*.  This is a convenience function which
   can be used from the module's initialization function.

>    On success, return ``0``. On error, raise an exception and return ``-1``.

Reviewer: Please keep this general statement. There are many cases where the function can fail. For example, if mod is not a module or has no dictionary. PyDict_SetItemString() can fail with MemoryError, etc.

   If *value* is ``NULL`` (indicating failure), raise an exception and return
   ``-1``.  If an exception is set, raise that exception.  If no exception is

>    set, raise :exc:`SystemError`.

Reviewer: This documentation sounds misleading to me. The normal case when value is NULL is that an exception is already set, so the function does **not** "raise an exception". Also, I dislike documenting the exact exception type for implementation details: please don't document "SystemError".

Reviewer 2: > please don't document "SystemError".

That was my fault; I encouraged it in https://github.com/python/cpython/pull/101957#discussion_r1111313045. Sorry 'bout that!

   On success, return ``0``. On error, raise an exception and return ``-1``.

   Return ``-1`` if *value* is ``NULL``. In this case, the function raises an

>    exception if called without an exception raised.

Reviewer 2: I find the second sentence a bit convoluted, but I think it is nice to mention that an exception can be raised. Perhaps @AlexWaygood has a good suggestion for a rewording? :)

Author: Thanks for the comments! 

I agree :) I played around with a few different versions but I couldn't find one that I was completely happy with TBH.

Reviewer 3: (I'm on vacation this week, but I'll try to take a look when I get back :)

Reviewer 2: Resolved as outdated; will start a new thread on the latest changes.

>    Return ``-1`` if *value* is ``NULL``. In this case, if the function is
>    called without an exception already set, it raises an exception.

Reviewer 2: This is better, but I still find it slightly convoluted. BTW, we should document which exception will be raised.

How about something like this:

```suggestion
   If *value* is ``NULL`` and and exception is set, return ``-1``.
   If *value* is ``NULL`` and no exception is set,
   :exc:`SystemError` is raised.
```

Author: Thanks!
- ``-1`` is returned in both cases, so that would need to be factored in.
- The reason I didn't specify the exception was because it wasn't done for the other cases, so I assumed maybe we didn't want to offer any guarantees.

How about the following?

```suggestion
   If *value* is ``NULL`` and an exception is set, return ``-1``. If *value*
   is ``NULL`` and no exception is set, raise :exc:`SystemError` and return
   ``-1``.
```

Reviewer 2: > `-1` is returned in both cases, so that would need to be factored in.

True. Perhaps the best candidate is the one proposed with commit 16c5efaf896eb27c836340c090122261c43eb84e

Author: I like the last suggestion TBH (derived directly from yours).  While longer, it is clearer and there's less room for misinterpretation. 

It explains logically exactly what it does in the case that *value* is ``NULL``.

I feel like as long as it does that, all should be good 😊.

>    On success, return ``0``. On error, raise an exception and return ``-1``.

Reviewer 2: Looks like the second sentence here has been made redundant.

Author: Hmm, yeah, I think the sentence you point out is, in general, confusing.

I think the message we want to get across is:

 - If the object is successfully added, return ``0``.
 - Otherwise, return ``-1``; in this case, an exception *may* be raised.

Then we would need to consider whether we want to enumerate all the possible exceptions, or leave that for the user to check. I'm kind of leaning towards the latter, although I think the *value*-is-`NULL` case should be described, especially because there are already examples for it in the docs.

What do you think?

Reviewer 2: > [...] or leave that for the user to check.

I'm -1 for that option; the reference should be succinct and exact (see [Diátaxis](https://diataxis.fr/reference/)).

I also think we're better off with removing the _"On error..." sentence above; we've covered the error case in your new and improved text below.

Reviewer 2: Marking as resolved; follow-up in https://github.com/python/cpython/pull/101957#discussion_r1175816245

   If *value* is ``NULL`` and an exception is set, return ``-1``. If *value*
   is ``NULL`` and no exception is set, raise :exc:`SystemError` and return

>    ``-1``.

Reviewer 3: It feels tricky to get the wording right here! I agree with @erlend-aasland that it feels like there's a _little_ bit of duplication currently.

How about something like this? Can't add a GitHub suggestion because the span includes deleted lines :(

```rst
   On success, return ``0``.

   If *value* is ``NULL`` (indicating failure), raise an exception and return
   ``-1``. If an exception is set, that exception will be raised. If no
   exception is set, :exc:`SystemError` will be raised.
```

Reviewer 2: cc. @OTheDev

Author: @erlend-aasland @AlexWaygood 
Sorry for the delay! I can get back to this in the weekend. Please feel free to take it over if it is urgent (:

Reviewer 2: No worries, no hurry :)

Author: Thanks for the suggestion, Alex!
